Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp

```diff
@@ -266,77 +257,45 @@ void CCloudyPanelPluginModule::StreamFrameToClient() {
 	// use VideoPipe (class variable) to pass frames to encoder
 	uint32 *PixelBuffer;
 	FColor Pixel;
-	for (int i = 0; i < NumberOfPlayers; i++) {
-		PixelBuffer = new uint32[sizeX * sizeY * PIXEL_SIZE];
-		//UE_LOG(ModuleLog, Warning, TEXT("FrameIndex %d"), FrameIndex);
-		for (int j = 0; j < FrameBufferList[i].Num(); j++) {
+	PixelBuffer = new uint32[sizeX * sizeY * PIXEL_SIZE];
+
+	for (int i = 0; i < NumberOfPlayers; i++) {	
+		int FrameSize = FrameBufferList[i].Num();
+		for (int j = 0; j < FrameSize; j++) {
 			
 			Pixel = FrameBufferList[PlayerFrameMapping[i]][j];
-			PixelBuffer[j] = Pixel.A * 256 * 256 * 256 + Pixel.B * 256 * 256 + Pixel.G * 256 + Pixel.R;
+			//PixelBuffer[j] = Pixel.A * 256 * 256 * 256 + Pixel.B * 256 * 256 + Pixel.G * 256 + Pixel.R;
 			// equivalent function using bitshift - can compare performance later
-			// PixelBuffer[i] = Pixel.A << 24 | Pixel.B << 16 | Pixel.G << 8 | Pixel.R;
+			PixelBuffer[j] = Pixel.A << 24 | Pixel.B << 16 | Pixel.G << 8 | Pixel.R;
 		}
 
-		fwrite(PixelBuffer, sizeX * PIXEL_SIZE/2, sizeY/2, VideoPipeList[i]);
-
-		delete[]PixelBuffer;
+		fwrite(PixelBuffer, halfSizeX * PIXEL_SIZE, halfSizeY, VideoPipeList[i]);
 	}
-
+	delete[]PixelBuffer;
 }
 
 
 // Split screen for 4 player
-void CCloudyPanelPluginModule::Split4Player(TArray<FColor> FrameBuffer, int FrameOffset) {
-
-	// empty buffers first
-	for (int i = 0; i < NumberOfPlayers; i++) {
-		FrameBufferList[i].Empty();
-	}
-
-	for (int i = 0; i < FrameBuffer.Num() / 2; i++) {
-		// Player 1
-		if (i % sizeX < sizeX / 2) { // potentially expensive. may have to optimise later
-			FrameBufferList[0].Add(FrameBuffer[i]);
-		}
-		else { // Player 2
-			if (NumberOfPlayers > 1)
-				FrameBufferList[1].Add(FrameBuffer[i]);
-		}
-	}
-
-	// deal with odd frame height
-	for (int i = FrameOffset; i < FrameBuffer.Num(); i++) {
-		if (i % sizeX < sizeX / 2 && NumberOfPlayers > 2) { // Player 3
-			FrameBufferList[2].Add(FrameBuffer[i]);
-		}
-		else { // Player 4
-			if (NumberOfPlayers > 3)
-				FrameBufferList[3].Add(FrameBuffer[i]);
-		}
-	}
+void CCloudyPanelPluginModule::Split4Player() {
 
-}
+	FViewport* ReadingViewport = GEngine->GameViewport->Viewport;
 
+	FIntRect Screen1 = FIntRect(0, 0, halfSizeX, halfSizeY);
+	FIntRect Screen2 = FIntRect(halfSizeX, 0, sizeX, halfSizeY);
+	FIntRect Screen3 = FIntRect(0, halfSizeY, halfSizeX, sizeY);
+	FIntRect Screen4 = FIntRect(halfSizeX, halfSizeY, sizeX, sizeY);
+	FReadSurfaceDataFlags flags = FReadSurfaceDataFlags(ERangeCompressionMode::RCM_MinMaxNorm,
+		ECubeFace::CubeFace_NegX);
```

Comment:
This function is being called once per frame. Can this initialization be shifted out of the function so it does not need to be done every frame?